File: core/src/main/java/arez/Arez.java
package arez;

import arez.component.Verifiable;
import grim.annotations.OmitClinit;
import grim.annotations.OmitSymbol;
import javax.annotation.Nonnull;
import org.realityforge.braincheck.BrainCheckConfig;
import static org.realityforge.braincheck.Guards.*;

/**
 * Provide access to an instance of ArezContext and Arez global configuration settings.
 * The {@link #context()} should always return the current context. If {@link #areZonesEnabled()}
 * is false it will return a singleton otherwise the appropriate context for the zone will be
 * invoked.
 */
@OmitClinit
public final class Arez
{
  private Arez()
  {
  }

  /**
   * Return true if zones are enabled, false otherwise.
   *
   * @return true if zones are enabled, false otherwise.
   */
  @OmitSymbol
  public static boolean areZonesEnabled()
  {
    return ArezConfig.areZonesEnabled();
  }

  /**
   * Return true if user should pass names into API methods, false if should pass null.
   *
   * @return true if user should pass names into API methods, false if should pass null.
   */
  @OmitSymbol
  public static boolean areNamesEnabled()
  {
    return ArezConfig.areNamesEnabled();
  }

  /**
   * Return true if {@link Verifiable} will verify components be used, false if not.
   *
   * @return true if {@link Verifiable} will verify components be used, false if not.
   */
  @OmitSymbol
  public static boolean isVerifyEnabled()
  {
    return ArezConfig.isVerifyEnabled();
  }

  /**
   * Return true if spies are enabled.
   *
   * @return true if spies are enabled, false otherwise.
   */
  @OmitSymbol
  public static boolean areSpiesEnabled()
  {
    /*
     * Spy's use debug names so we can not enable spies without names.
     */
    return areNamesEnabled() && ArezConfig.areSpiesEnabled();
  }

  /**
   * Return true if references are enabled.
   *
   * @return true if references are enabled, false otherwise.
   */
  @OmitSymbol
  public static boolean areReferencesEnabled()
  {
    return ArezConfig.areReferencesEnabled();
  }

  /**
   * Return true if observable properties, computable properties or query results that are of type collection are wrapped in unmodifiable variant prior to returning.
   *
   * @return true if observable properties, computable properties or query results that are of type collection are wrapped in unmodifiable variant prior to returning.
   */
  @OmitSymbol
  public static boolean areCollectionsPropertiesUnmodifiable()
  {
    return ArezConfig.areCollectionsPropertiesUnmodifiable();
  }

  /**
   * Return true if property introspectors for Observables are enabled.
   *
   * @return true if property introspectors for Observables are enabled, false otherwise.
   */
  @OmitSymbol
  public static boolean arePropertyIntrospectorsEnabled()
  {
    return areSpiesEnabled() && ArezConfig.arePropertyIntrospectorsEnabled();
  }

  /**
   * Return true if registries for top level reactive components are enabled.
   *
   * @return true if registries for top level reactive components are enabled, false otherwise.
   */
  @OmitSymbol
  public static boolean areRegistriesEnabled()
  {
    return areNamesEnabled() && ArezConfig.areRegistriesEnabled();
  }

  /**
   * Return true if native components are enabled.
   *
   * @return true if native components are enabled, false otherwise.
   */
  @OmitSymbol
  public static boolean areNativeComponentsEnabled()
  {
    return ArezConfig.areNativeComponentsEnabled();
  }

  /**
   * Return true if observer error handlers are enabled.
   *
   * @return true if observer error handlers are enabled, false otherwise.
   */
  @OmitSymbol
  public static boolean areObserverErrorHandlersEnabled()
  {
    return ArezConfig.areObserverErrorHandlersEnabled();
  }

  /**
   * Return true if Arez should enforce transaction modes.
   *
   * @return true if Arez should enforce transaction modes.
   */
  @OmitSymbol
  static boolean shouldEnforceTransactionType()
  {
    return ArezConfig.enforceTransactionType();
  }

  /**
   * Return true if a task interceptor can be specified.
   *
   * @return true if a task interceptor can be specified
   */
  @OmitSymbol
  public static boolean isTaskInterceptorEnabled()
  {
    return ArezConfig.isTaskInterceptorEnabled();
  }

  /**
   * Return true if invariants will be checked.
   *
   * @return true if invariants will be checked.
   */
  @OmitSymbol
  public static boolean shouldCheckInvariants()
  {
    return ArezConfig.checkInvariants() && BrainCheckConfig.checkInvariants();
  }

  /**
   * Return true if expensive invariants will be checked.
   *
   * @return true if expensive invariants will be checked.
   */
  @OmitSymbol
  public static boolean shouldCheckExpensiveInvariants()
  {
    return ArezConfig.checkExpensiveInvariants() && BrainCheckConfig.checkInvariants();
  }

  /**
   * Return true if apiInvariants will be checked.
   *
   * @return true if apiInvariants will be checked.
   */
  @OmitSymbol
  public static boolean shouldCheckApiInvariants()
  {
    return ArezConfig.checkApiInvariants() && BrainCheckConfig.checkApiInvariants();
  }

  /**
   * Return true if active tasks will be purged if the scheduler is still running after 100 rounds.
   *
   * @return true if active tasks will be purged if the scheduler is still running after 100 rounds.
   */
  @OmitSymbol
  public static boolean purgeTasksWhenRunawayDetected()
  {
    return ArezConfig.purgeTasksWhenRunawayDetected();
  }

  /**
   * Return the ArezContext from the provider.
   *
   * @return the ArezContext.
   */
  @Nonnull
  public static ArezContext context()
  {
    return areZonesEnabled() ? ZoneHolder.context() : ArezContextHolder.context();
  }

  /**
   * Create a new zone.
   * This zone is not yet activated.
   *
   * @return the new zone.
   */
  @OmitSymbol( unless = "arez.enable_zones" )
  @Nonnull
  public static Zone createZone()
  {
    if ( shouldCheckApiInvariants() )
    {
      apiInvariant( Arez::areZonesEnabled, () -> "Arez-0001: Invoked Arez.createZone() but zones are not enabled." );
    }
    return new Zone();
  }

  /**
   * Save the old zone and make the specified zone the current zone.
   */
  @OmitSymbol( unless = "arez.enable_zones" )
  static void activateZone( @Nonnull final Zone zone )
  {
    ZoneHolder.activateZone( zone );
  }

  /**
   * Restore the old zone.
   * This takes the zone that was current when {@link #activateZone(Zone)} was called for the active zone
   * and restores it to being the current zone.
   */
  @OmitSymbol( unless = "arez.enable_zones" )
  static void deactivateZone( @Nonnull final Zone zone )
  {
    ZoneHolder.deactivateZone( zone );
  }

  /**
   * Return the current zone.
   *
   * @return the current zone.
   */
  @OmitSymbol( unless = "arez.enable_zones" )
  @Nonnull
  static Zone currentZone()
  {
    return ZoneHolder.currentZone();
  }
}

File: core/src/main/java/arez/arez.js
/**
 * This file provides the @defines for arez configuration options.
 * See ArezConfig.java for details.
 */
goog.provide('arez');
goog.require('jre');

/** @define {string} */
arez.environment = goog.define('arez.environment', 'production');
goog.module.get('jre').addSystemPropertyFromGoogDefine('arez.environment', arez.environment);

/** @define {string} */
arez.enable_names = goog.define('arez.enable_names', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine('arez.enable_names', arez.enable_names);

/** @define {string} */
arez.enable_references = goog.define('arez.enable_references', 'true');
goog.module.get('jre').addSystemPropertyFromGoogDefine('arez.enable_references', arez.enable_references);

/** @define {string} */
arez.enable_property_introspection = goog.define('arez.enable_property_introspection', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine(
  'arez.enable_property_introspection',
  arez.enable_property_introspection);

/** @define {string} */
arez.enable_verify = goog.define('arez.enable_verify', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine('arez.enable_verify', arez.enable_verify);

/** @define {string} */
arez.enforce_transaction_type = goog.define('arez.enforce_transaction_type', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine(
  'arez.enforce_transaction_type',
  arez.enforce_transaction_type);

/** @define {string} */
arez.purge_tasks_when_runaway_detected = goog.define('arez.purge_tasks_when_runaway_detected', 'true');
goog.module.get('jre').addSystemPropertyFromGoogDefine(
  'arez.purge_tasks_when_runaway_detected',
  arez.purge_tasks_when_runaway_detected);

/** @define {string} */
arez.collections_properties_unmodifiable = goog.define('arez.collections_properties_unmodifiable', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine(
  'arez.collections_properties_unmodifiable',
  arez.collections_properties_unmodifiable);

/** @define {string} */
arez.enable_zones = goog.define('arez.enable_zones', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine('arez.enable_zones', arez.enable_zones);

/** @define {string} */
arez.enable_spies = goog.define('arez.enable_spies', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine('arez.enable_spies', arez.enable_spies);

/** @define {string} */
arez.enable_native_components = goog.define('arez.enable_native_components', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine(
  'arez.enable_native_components',
  arez.enable_native_components);

/** @define {string} */
arez.enable_registries = goog.define('arez.enable_registries', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine('arez.enable_registries', arez.enable_registries);

/** @define {string} */
arez.enable_observer_error_handlers = goog.define('arez.enable_observer_error_handlers', 'true');
goog.module.get('jre').addSystemPropertyFromGoogDefine(
  'arez.enable_observer_error_handlers',
  arez.enable_observer_error_handlers);

/** @define {string} */
arez.enable_task_interceptor = goog.define('arez.enable_task_interceptor', 'true');
goog.module.get('jre').addSystemPropertyFromGoogDefine(
  'arez.enable_task_interceptor',
  arez.enable_task_interceptor);

/** @define {string} */
arez.check_invariants = goog.define('arez.check_invariants', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine('arez.check_invariants', arez.check_invariants);

/** @define {string} */
arez.check_expensive_invariants = goog.define('arez.check_expensive_invariants', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine(
  'arez.check_expensive_invariants',
  arez.check_expensive_invariants);

/** @define {string} */
arez.check_api_invariants = goog.define('arez.check_api_invariants', 'false');
goog.module.get('jre').addSystemPropertyFromGoogDefine(
  'arez.check_api_invariants',
  arez.check_api_invariants);

/** @define {string} */
arez.logger = goog.define('arez.logger', 'none');
goog.module.get('jre').addSystemPropertyFromGoogDefine('arez.logger', arez.logger);
